README.md
[![Build and Test](https://github.com/csailer/chux-datastore/actions/workflows/build_and_test.yml/badge.svg?branch=master)](https://github.com/chuxorg/chux-datastore/actions/workflows/build_and_test.yml)
[![Go Report Card](https://goreportcard.com/badge/github.com/chuxorg/chux-datastore)](https://goreportcard.com/report/github.com/chuxorg/chux-datastore)
# Chux Datastore Go Library - chux-datastore

chux-datastore is a simple and easy-to-use Go library for performing basic CRUD operations on MongoDB. The library provides a clean and straightforward interface for connecting to MongoDB, as well as creating, retrieving, updating, and deleting documents.

## Driving Forces

Ah, the treacherous realm of software development! A chaotic wasteland where arbitrary deadlines circle like vultures, ever ready to swoop down upon the unsuspecting developer from the fantastical realms of project plans, conjured by oxygen thieves bearing titles like Project Manager and Scrum Master. Amidst this silly world, one must conquer the ferocious beast of repetitive code. But fret not, for chux-mongo rides to the rescue, charging heroically into the melee!

Envision this: You are a Software Engineer with A.D.D. battling the banal and the humdrum of avoiding the "Crotch Punching Gnomes" that lurk beneath your desk and ponce, without warning, at every unit test failure. One fateful day, your gaze falls upon your jumbled codebase, only to discover that the ominous MongoDB CRUD code is the recurring nemesis in this grand saga. True, it functions, but its unsightly presence is a thorn in your side. If only something could emerge to liberate you from this ordeal...

Enter `chux-mongo`, a formidable Excalibur forged amidst the inferno of ingenuity and practicality. In one swift motion, it beheads the monstrous hydra of boilerplate, emancipating you from its tyrannical clutches. With chux-mongo as your steadfast ally, your codebase transforms into a pristine sanctuary where CRUD operations frolic freely like untamed mustangs.

And what of the days yet to come? Shall this valiant library withstand the sands of time, lending its might to other endeavors in their pursuit of greatness? The answer, dear compatriot, resounds with a thunderous affirmation. For chux-mongo transcends the fleeting present, safeguarding the myriad services that shall voyage through the perilous landscape of reading and writing to and from Mongo for countless eons.

In closing, chux-mongo emerges as the mythical hero we never knew we yearned for, restoring harmony to the pandemonium, severing the chains of monotony, and banishing the specter of boilerplate Mongo code to the annals of distant memory. 

The Repos are Dark and Full of Terrors.

Nah, Seriously. Very simple library that encapsulates CRUD operations for [Mongo DB](https://www.mongodb.com) with Golang >= 1.19.
## Features
- Simple MongoDB connection setup
- Basic CRUD operations: Create, Read, Update, and Delete
- Supports custom MongoDB document structures
## Getting Started

**Prerequisites**
- Go 1.19 or later
- MongoDB server
- Installation

To install chux-datastore, run the following command:

```sh
go get github.com/chuxorg/chux-datastore
```

## Usage
This example demonstrates how to perform CRUD operations using the provided code with a 
custom struct called `MyMongoDocument` that implements the `IMongoDocument`interface. 
The `main()` function initializes a `MongoDB` instance, and then it creates, retrieves, 
updates, and deletes a document in a `MongoDB` collection.

First, download `chux-mongo`:

```shell
$ go get github.com/chuxorg/chux-datastore
```

```go
package main

import (
	"fmt"
	"log"

	"github.com/chuxorg/chux-datastore/db"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Define a custom struct that implements the IMongoDocument interface
type MyMongoDocument struct {
	ID        primitive.ObjectID `bson:"_id,omitempty"`
	FirstName string             `bson:"firstName"`
	LastName  string             `bson:"lastName"`
}

func (m *MyMongoDocument) GetCollectionName() string {
	return "myCollection"
}

func (m *MyMongoDocument) GetDatabaseName() string {
	return "myDatabase"
}

func (m *MyMongoDocument) GetURI() string {
	return "mongodb://localhost:27017"
}

func (m *MyMongoDocument) GetID() primitive.ObjectID {
	return m.ID
}

func main() {
	// Initialize the MongoDB instance
	mongoDB := db.New(
		db.WithURI("mongodb://localhost:27017"),
		db.WithTimeout(30),
	)

	// Create a new document
	doc := &MyMongoDocument{
		FirstName: "John",
		LastName:  "Doe",
	}
	err := mongoDB.Create(doc)
	if err != nil {
		log.Fatalf("Error creating document: %v", err)
	}
	fmt.Printf("Created document: %+v\n", doc)

	// Get document by ID
	foundDoc := &MyMongoDocument{}
	id := doc.ID.Hex()
	_, err = mongoDB.GetByID(foundDoc, id)
	if err != nil {
		log.Fatalf("Error getting document by ID: %v", err)
	}
	fmt.Printf("Found document by ID: %+v\n", foundDoc)

	// Get all documents
	allDocs, err := mongoDB.GetAll(&MyMongoDocument{})
	if err != nil {
		log.Fatalf("Error getting all documents: %v", err)
	}
	fmt.Println("All documents:")
	for _, d := range allDocs {
		fmt.Printf("%+v\n", d)
	}

	// Update a document
	updatedDoc := &MyMongoDocument{
		ID:        doc.ID,
		FirstName: "Jane",
		LastName:  "Doe",
	}
	err = mongoDB.Update(updatedDoc, id)
	if err != nil {
		log.Fatalf("Error updating document: %v", err)
	}
	fmt.Printf("Updated document: %+v\n", updatedDoc)

	// Delete a document
	err = mongoDB.Delete(updatedDoc, id)
	if err != nil {
		log.Fatalf("Error deleting document: %v", err)
	}
	fmt.Println("Deleted document")
}


```

# Makefile

- `make test` - Runs all tests in `chux-mongo`.
- `make test-release` - Commits, tags, and releases `chux-mongo`  
   &nbsp; 
   To release and version, pass in major and minor values on the command line. If either major or minor has a value, the patch number is set to zero. If neither major nor minor has a value, the patch number is incremented by 1.
   You can run the target with or without the MAJOR_VALUE and MINOR_VALUE variables, like this:
   &nbsp; 
   To bump the patch version:
   ```shell
   make release-version
   ```
   &nbsp; 
   To set a new major version and reset minor and patch to zero:
   ```shell
   MAJOR_VALUE=2 make release-version
   ```
   &nbsp; 
   To set a new minor version and reset patch to zero:
   ```shell
   MINOR_VALUE=3 make release-version 
   ```
   &nbsp; 
   To set both new major and minor versions and reset patch to zero:
   ```shell
   MAJOR_VALUE=2 MINOR_VALUE=3 make release-version 
   ``` 
## Unit Tests
The unit tests use an in-memory MongoDB from Ben Weissmann named [memongo](https://github.com/benweissmann/memongo).
I really like it and, to me, it beats Mocking MongoDB. It works very well locally but I haven't had much luck in GitHub
Pipelines.
## Contributing
Contributions are welcome! Please feel free to submit issues and pull requests.

## Projects Using this Library

- [chux-models](github.com/chuxorg/chux-models)

## License
chux-mongo is released under the [GNU General Public License v3](https://www.gnu.org/licenses/gpl-3.0.en.html)


CHANGELOG.md
# Changelog

## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG


## 3460TAG



db/mongo_test.go
package db

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

var mongoDB MongoDB

// TestNew tests the New function of the MongoDB struct
func TestNew(t *testing.T) {
	uri := "mongodb://localhost:27017"
	timeout := 30.0
	databaseName := "test"
	collectionName := "test"

	mongoDB = New(
		WithURI(uri),
		WithTimeout(timeout),
		WithDatabaseName(databaseName),
		WithCollectionName(collectionName),
	)

	assert.Equal(t, uri, mongoDB.URI)
	assert.Equal(t, timeout, mongoDB.Timeout)
	assert.Equal(t, databaseName, mongoDB.DatabaseName)
	assert.Equal(t, collectionName, mongoDB.CollectionName)
}


config/config.go
// Package: config
// Description: This file contains the code to load the configuration file
// and unmarshal it into a struct.
// Author: Chuck Sailer
// Date: 2023-04-07
// Version: 1.0.0
// License: gpl-3.0
// Copyright: Chuck Sailer
// Credits: [Chuck Sailer]
// Maintainer: Chuck Sailer
// Status: Production
package config

import (
	"fmt"

	"github.com/spf13/viper"
)

// DataStoreConfig is the struct that holds the configuration
// for the data store.
// The struct is populated by the configuration file.
// The struct is used to initialize the data store.
// The struct is also used to initialize the logger.
//
// The configuration file is in YAML format.
//
// The DataStoreConfig is provided so chux-datastore can be used
// as a a stand-alone library. However, the typical use case is
// to use chux-datastore as a library in a larger application.
// In that case, the larger application will have its own
// configuration file. The larger application will use the
// chux-datastore library to initialize the data store.
// The larger application will also use the chux-datastore
// library to initialize the logger.
//
// In these cases, the larger application will have its own
// configuration file and will pass the configuration section
// to chux-datastore.
type DataStoreConfig struct {
	Logging struct {
		Level string `mapstructure:"level"`
	} `mapstructure:"logging"`
	DataStores []struct {
		DataStore struct {
			Mongo struct {
				URI            string `mapstructure:"uri"`
				Timeout        int    `mapstructure:"timeout"`
				DatabaseName   string `mapstructure:"databaseName"`
				CollectionName string `mapstructure:"collectionName"`
			} `mapstructure:"mongo"`
			Redis struct {
				URI string `mapstructure:"uri"`
			} `mapstructure:"redis"`
		} `mapstructure:"dataStore"`
	} `mapstructure:"DataStores"`
}

func New(options ...func(*DataStoreConfig)) *DataStoreConfig {

	dsc := &DataStoreConfig{}
	for _, o := range options {
		o(dsc)
	}
	return dsc 
}

func WithDataStoreConfig(dsc *DataStoreConfig) func(*DataStoreConfig) {
	return func(cfg *DataStoreConfig) {
		cfg = dsc
	}
}


func LoadConfig(configPath string) (*DataStoreConfig, error) {
	v := viper.New()

	// Set the configuration file format and name
	v.SetConfigType("yaml")
	v.SetConfigName("config")

	// Add the configuration file path
	v.AddConfigPath(configPath)

	// Read the configuration file
	err := v.ReadInConfig()
	if err != nil {
		return nil, fmt.Errorf("failed to read the configuration file: %w", err)
	}

	// Unmarshal the configuration into the Config struct
	var cfg DataStoreConfig
	err = v.Unmarshal(&cfg)
	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal the configuration: %w", err)
	}

	return &cfg, nil
}


errors/errors.go
package errors

import "fmt"

// ChuxDataStoreError is a custom error type
// that wraps an error and adds a message
// to the error.
// This is the error that is returned by
// all functions in chux-datastore that return
// an error.
type ChuxDataStoreError struct {
	// Message is the message that is
	// given by chux-parser when an error
	// occurs.
	// This message is used to provide
	// more context to the error.
	// The Err field contains the actual 
	// error that occurred.
	Message string
	Err     error
}

// NewChuxParserError returns a new ChuxDataStoreError
func NewChuxDataStoreError(message string, err error) *ChuxDataStoreError {
	return &ChuxDataStoreError{
		Message: message,
		Err:     err,
	}
}

func (e *ChuxDataStoreError) Error() string {
	return e.Message
}

// Unwrap returns the underlying error without
// the message added by chux-parser.
func (e *ChuxDataStoreError) Unwrap() error {
	return e.Err
}


// handleError is a helper function that handles
// errors occuring in chux-parser. This means 
// that it prints the error message and the
// underlying error. It will also log the error	
func handleError(err error) {
	fmt.Printf("Error: %v\n", err)
}

logging/fileLogger.go
package logging

import (
	"archive/zip"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"time"
)
type FileLogger struct {
	file          *os.File
    maxFileSize   int64
    logDirectory  string
    logFilePrefix string
	// The logChan is used to send log messages to the logger goroutine.
	logChan       chan logEntry
	// The quitChan is used to send a signal to the logger goroutine to stop.
	quitChan      chan bool
}

type logEntry struct {
    level string
    msg   string
    args  []interface{}
}

func New(logDirectory, logFilePrefix string, maxFileSize int64) (*FileLogger, error) {
    if err := os.MkdirAll(logDirectory, 0755); err != nil {
        return nil, err
    }

    logFile, err := createLogFile(logDirectory)
    if err != nil {
        return nil, err
    }

    fileLogger := FileLogger{
        file:          logFile,
        maxFileSize:   maxFileSize,
        logDirectory:  logDirectory,
        logFilePrefix: logFilePrefix,
		logChan:       make(chan logEntry, 100),
        quitChan:      make(chan bool),
    }
	go fileLogger.logHandler()

    return &fileLogger, nil
}

// implement ILogger interface
func (fl *FileLogger) Debug(msg string, args ...interface{}) {
	fl.log("DEBUG", msg, args...)
}

func (fl *FileLogger) Info(msg string, args ...interface{}) {
	fl.log("INFO", msg, args...)
}

func (fl *FileLogger) Warn(msg string, args ...interface{}) {
	fl.log("WARN", msg, args...)
}

func (fl *FileLogger) Error(msg string, args ...interface{}) {
	fl.log("ERROR", msg, args...)
}

// createLogFile creates a new log file in the given directory with the given prefix.
func createLogFile(logDirectory string) (*os.File, error) {
    logFilePrefix := "chux-datastore-log"
	timestamp := time.Now().Format("2006-01-02")
    fileName := fmt.Sprintf("%s-%s.log", logFilePrefix, timestamp)
    filePath := filepath.Join(logDirectory, fileName)
    return os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
}

func (fl *FileLogger) log(level, msg string, args ...interface{}) {
    fl.logChan <- logEntry{level: level, msg: msg, args: args}
}

func (fl *FileLogger) needsRotation() bool {
    fileInfo, err := fl.file.Stat()
    if err != nil {
        return false
    }
    return fileInfo.Size() >= fl.maxFileSize
}

func (fl *FileLogger) rotate() {
    // Close the current log file
    fl.file.Close()

    // Zip the old log file
    oldFilePath := fl.file.Name()
    zipFilePath := oldFilePath + ".zip"
    if err := zipFile(oldFilePath, zipFilePath); err != nil {
        fmt.Printf("Error zipping old log file: %v\n", err)
    } else {
        os.Remove(oldFilePath)
    }

    // Create a new log file
    newLogFile, err := createLogFile(fl.logDirectory)
    if err != nil {
        fmt.Printf("Error creating new log file: %v\n", err)
        return
    }
    fl.file = newLogFile
}

func (fl *FileLogger) logHandler() {
    for {
        select {
        case entry := <-fl.logChan:
            // Check if the file size exceeds the maximum size
            if fl.needsRotation() {
                fl.rotate()
            }

            log.SetOutput(fl.file)
            log.Printf("%s: %s", entry.level, fmt.Sprintf(entry.msg, entry.args...))
        case <-fl.quitChan:
            return
        }
    }
}

func (fl *FileLogger) Close() {
    fl.quitChan <- true
    close(fl.logChan)
    fl.file.Close()
}


// zipFile zips a log file and returns an error if any
// occurs during the process.
func zipFile(source, target string) error {
    zipFile, err := os.Create(target)
    if err != nil {
        return err
    }
    defer zipFile.Close()

    archive := zip.NewWriter(zipFile)
    defer archive.Close()

    sourceFile, err := os.Open(source)
    if err != nil {
        return err
    }
    defer sourceFile.Close()

    fileInfo, err := sourceFile.Stat()
    if err != nil {
        return err
    }

    header, err := zip.FileInfoHeader(fileInfo)
    if err != nil {
        return err
    }

    // Use the file name as the zip entry name
    header.Name = filepath.Base(source)

    // Set the compression method to Deflate
    header.Method = zip.Deflate

    writer, err := archive.CreateHeader(header)
    if err != nil {
        return err
    }

    _, err = io.Copy(writer, sourceFile)
    return err
}

db/mongo.go
// db package
package db

import (
	"context"
	"fmt"
	"log"
	"reflect"
	"time"

	"github.com/chuxorg/chux-datastore/errors"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
)

// The IMongoDocument interface is used to define the methods that must be implemented by a struct that is used to store
// a Mongo Document. The interface is used to allow for a single struct to be used for multiple collections. The struct
// can implement the interface and override the default collection and database name with its own implementation of the
// interface methods.
// Example:
//
//		type MyMongoDocument struct {
//				ID        primitive.ObjectID `bson:"_id,omitempty"`
//				FirstName string             `bson:"firstName"`
//				LastName  string             `bson:"lastName"`
//		}
//
//		func (m *MyMongoDocument) GetCollectionName() string {
//			return "myCollection"
//		}
//	 func (m *MyMongoDocument) GetDatabaseName() string {
//			return "myDatabase"
//		}
//	 func (m *MyMongoDocument) GetURI() string {
//			Use this if you want to override the default URI
//			return "mongodb://localhost:27017"
//		}

//go:generate mockery --name MongoDocument
type IMongoDocument interface {
	GetCollectionName() string
	GetDatabaseName() string
	GetURI() string
	GetID() primitive.ObjectID
	
}

type IMongoClientMethods interface {
	Connect(ctx context.Context) error
	Disconnect(ctx context.Context) error
	Database(name string) *mongo.Database
	Ping(ctx context.Context, rp *readpref.ReadPref) error
}

//go:generate mockery --name MongoClient
type IMongoClient interface {
	IMongoClientMethods
}

//go:generate mockery --name MongoDB
type IMongoDB interface {
	Connect() (*mongo.Client, error)
	Create(doc IMongoDocument, id string) error
	Delete(doc IMongoDocument) error
	GetByID(doc IMongoDocument, id string) error
	GetAll(doc IMongoDocument) ([]IMongoDocument, error)
	Update(doc IMongoDocument, id string) error
	CreateIndicies(fieldNames ...interface{}) (bool, error)
}

// The MongoDB struct is used to store the MongoDB configuration
type MongoDB struct {
	ID             primitive.ObjectID
	CollectionName string
	DatabaseName   string
	URI            string
	Timeout        float64
	_client        IMongoClient
}

// The _client variable is used to store the MongoDB client
var _client *mongo.Client

// The New func constructs the MongoDB struct with the given options.
// Example:
//
//	mongoDB := New(
//		WithURI("mongodb://localhost:27017"),
//		WithTimeout(30),
//		WithDatabaseName("test"),
//		WithCollectionName("test"),
//	)
func New(options ...func(*MongoDB)) *MongoDB {

	mdb := &MongoDB{}
	for _, o := range options {
		o(mdb)
	}
	return mdb
}

// WithURI is a functional option that sets the MongoDB URI.
//
// Example:
//
//	mongoDB := New(
//		WithURI("mongodb://localhost:27017"),
//	)
func WithURI(uri string) func(*MongoDB) {
	return func(s *MongoDB) {
		s.URI = uri
	}
}

// WithTimeout is a functional option that sets the MongoDB Timeout.
//
// Example:
//
//	mongoDB := New(
//		withTimeout(30),
//	)
func WithTimeout(timeout float64) func(*MongoDB) {
	return func(s *MongoDB) {
		s.Timeout = timeout
	}
}

// WithDatabaseName is a functional option that sets the MongoDB Database Name.
// Example:
//
//	mongoDB := New(
//		withDatabaseName("test"),
//	)
func WithDatabaseName(databaseName string) func(*MongoDB) {
	return func(s *MongoDB) {
		s.DatabaseName = databaseName
	}
}

// WithCollectionName is a functional option that sets the MongoDB Collection Name.
// Example:
//
//	mongoDB := New(
//		withCollectionName("test"),
//	)
func WithCollectionName(collectionName string) func(*MongoDB) {
	return func(s *MongoDB) {
		s.CollectionName = collectionName
	}
}

// The GetID() method is used to return the ID of the MongoDB struct.
func (m *MongoDB) GetID() primitive.ObjectID {
	return m.ID
}

// The Connect func is used to connect to the MongoDB server. It returns a mongo.Client and an error.
// Example:
//
//	mongoDB, err := Connect()
//	if err != nil {
//		return err
//	}
func (m *MongoDB) Connect() (*mongo.Client, error) {
	if _client != nil {
		// Client has already been created. Return it
		return _client, nil
	}
	if m.Timeout == 0 {
		m.Timeout = 30 // default value
	}
	timeoutDuration := time.Duration(m.Timeout) * time.Second
	if m.Timeout == 0 {
		timeoutDuration = 30 * time.Second // default value
	}

	var err error
	var uri string
	// Check the URI
	if len(m.URI) == 0 {
		// Set the uri to a default value
		uri = "mongodb://localhost:27017"
	} else {
		uri = m.URI
	}
	clientOptions := options.Client().
		ApplyURI(uri).
		SetConnectTimeout(timeoutDuration).        // Increase connection timeout
		SetServerSelectionTimeout(timeoutDuration) // Increase server selection timeout

	_client, err = mongo.NewClient(clientOptions)
	if err != nil {
		log.Fatal(err)
		return nil, NewChuxMongoError(fmt.Sprintf("Did not create mongo client for %s. . Check the inner error for details", m.URI), 1000, err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeoutDuration) // Increase context timeout
	defer cancel()

	err = _client.Connect(ctx)
	if err != nil {
		log.Fatal(err)
		return nil, NewChuxMongoError(fmt.Sprintf("Did not connect mongo client for %s. Check the inner error for details.", m.URI), 1001, err)
	}

	return _client, nil
}

// Creates a Mongo Document in the configured Mongo DB
// Example:
//
//	mongoDB := New()
//	err := mongoDB.Create(&MyMongoDocument{
//		FirstName: "John",
//		LastName:  "Doe",
//	})
func (m *MongoDB) Create(doc IMongoDocument) error {

	// Get the collection and insert then document
	collection, err := m.getCollection(doc)
	if err != nil {
		return err
	}
	// Create a context with a timeout of 10 seconds
	if m.Timeout == 0 {
		m.Timeout = 30 // default value
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(m.Timeout)*time.Second)
	defer cancel()

	// Insert the document
	insertResult, err := collection.InsertOne(ctx, doc)
	if err != nil {
		log.Fatal(err)
	}

	// Update the ID field of the document with the inserted ID
	if oid, ok := insertResult.InsertedID.(primitive.ObjectID); ok {
		val := reflect.ValueOf(doc).Elem()
		idField := val.FieldByName("ID")
		if idField.IsValid() && idField.CanSet() {
			idField.Set(reflect.ValueOf(oid))
		}
	}

	fmt.Println("Inserted document with ID:", insertResult.InsertedID)
	return nil
}

// Returns a Mongo Document by its ID from the configured Mongo DB
func (m *MongoDB) GetByID(doc IMongoDocument, id string) (interface{}, error) {
	client, err := m.Connect()
	if err != nil {
		return nil, err
	}

	collection := client.Database(doc.GetDatabaseName()).Collection(doc.GetCollectionName())
	objectID, err := primitive.ObjectIDFromHex(id)

	if err != nil {
		return nil, NewChuxMongoError("GetByID() Failed to Get ObjectIDFromHex. Check the inner error.", 1003, err)
	}
	err = collection.FindOne(context.Background(), bson.M{"_id": objectID}).Decode(doc)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			return nil, NewChuxMongoError("Document not found.", 1003, err)
		}
		return nil, NewChuxMongoError("GetByID failed. Check the inner error.", 1003, err)
	}
	return doc, nil
}

// Query allows for variadic parameters to query any field with any value of any type from MongoDB
// Example:
//
//	mongoDB := New()
//	docs, err := mongoDB.Query(&MyMongoDocument{}, "firstName", "John", "lastName", "Doe")
//	if err != nil {
//		return err
//	}
//	for _, doc := range docs {
//		fmt.Println(doc)
//	}
func (m *MongoDB) Query(doc IMongoDocument, queries ...interface{}) ([]IMongoDocument, error) {

	// prepare an empty slice to return in case there are no results
	emptySlice := make([]IMongoDocument, 0)
	// Check if the number of arguments is even (key-value pairs)
	if len(queries)%2 != 0 {
		return nil, NewChuxMongoError("Query() requires an even number of arguments for key-value pairs.", 1006, nil)
	}

	// Connect to the MongoDB client
	client, err := m.Connect()
	if err != nil {
		return nil, NewChuxMongoError("Query() error occurred connecting to Mongo", 1006, err)
	}

	// Get the collection from the specified database and collection names
	collection := client.Database(doc.GetDatabaseName()).Collection(doc.GetCollectionName())

	// Initialize the filter bson.M (a map) for MongoDB filtering
	filter := bson.M{}

	// Loop through the provided query parameters, in key-value pairs
	for i := 0; i < len(queries); i += 2 {
		// Cast the key to a string and check if the casting was successful
		key, ok := queries[i].(string)
		if !ok {
			return nil, NewChuxMongoError("Query() expects keys to be of type string.", 1006, nil)
		}
		// Add the key-value pair to the filter map
		filter[key] = queries[i+1]
	}

	// Execute the Find operation on the collection with the filter
	cursor, err := collection.Find(context.Background(), filter)
	if err != nil {
		// The query returned no results
		return emptySlice, nil
	}

	// Close the cursor when the function is done
	defer cursor.Close(context.Background())

	// Initialize a slice to store the decoded documents
	var docs []IMongoDocument

	// Loop through the cursor while there are more documents to fetch
	for cursor.Next(context.Background()) {
		// Create a new document instance based on the type of the provided doc
		newDoc := reflect.New(reflect.TypeOf(doc).Elem()).Interface().(IMongoDocument)

		// Decode the document from the cursor into the new document instance
		err := cursor.Decode(newDoc)
		if err != nil {
			return nil, NewChuxMongoError("Query() Failed to decode document. Check the inner error.", 1006, err)
		}

		// Append the new document to the docs slice
		docs = append(docs, newDoc)
	}

	// Check for errors in the cursor
	if err := cursor.Err(); err != nil {
		return nil, NewChuxMongoError("Query() Cursor error. Check the inner error.", 1006, err)
	}

	// If no documents were found, return an error
	if len(docs) == 0 {
		return emptySlice, nil
	}

	// Return the documents slice and a nil error
	return docs, nil
}

// Returns all Mongo Documents from the configured Mongo DB
// Example:
//
//	mongoDB := New()
//	docs, err := mongoDB.GetAll(&MyMongoDocument{})
//	if err != nil {
//		return err
//	}
//	for _, doc := range docs {
//		fmt.Println(doc)
//	}
func (m *MongoDB) GetAll(doc IMongoDocument) ([]IMongoDocument, error) {
	client, err := m.Connect()
	if err != nil {
		return nil, err
	}

	collection := client.Database(doc.GetDatabaseName()).Collection(doc.GetCollectionName())

	cursor, err := collection.Find(context.Background(), bson.M{})
	if err != nil {
		return nil, NewChuxMongoError("GetAll() Failed to find documents. Check the inner error.", 1004, err)
	}

	defer cursor.Close(context.Background())

	var docs []IMongoDocument
	for cursor.Next(context.Background()) {
		newDoc := reflect.New(reflect.TypeOf(doc).Elem()).Interface().(IMongoDocument)
		err := cursor.Decode(newDoc)
		if err != nil {
			return nil, NewChuxMongoError("GetAll() Failed to decode document. Check the inner error.", 1004, err)
		}
		docs = append(docs, newDoc)
	}

	if err := cursor.Err(); err != nil {
		return nil, NewChuxMongoError("GetAll() Cursor error. Check the inner error.", 1004, err)
	}

	if len(docs) == 0 {
		return nil, NewChuxMongoError("No documents found.", 1004, nil)
	}

	return docs, nil
}

// Updates a Mongo Document by its ID from the configured Mongo DB
// Example:
//
//	mongoDB := New(
//		WithDatabaseName("mydb"),
//		WithCollectionName("mycollection"),
//		WithURI("mongodb://localhost:27017"),
//	)
//	err := mongoDB.Update(&MyMongoDocument{
//		FirstName: "John",
//		LastName:  "Doe",
//	}, "5e9b9b9b9b9b9b9b9b9b9b9b")
func (m *MongoDB) Update(doc IMongoDocument, id string) error {
	client, err := m.Connect()

	if err != nil {
		return err
	}

	collection := client.Database(doc.GetDatabaseName()).Collection(doc.GetCollectionName())
	objectID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		log.Fatal(err)
		return NewChuxMongoError("Update() Failed to Get ObjectIDFromHex. Check the inner error.", 1004, err)
	}
	update := bson.M{
		"$set": doc,
	}
	result, err := collection.UpdateOne(context.Background(), bson.M{"_id": objectID}, update)
	if err != nil {
		log.Fatal(err)
		return NewChuxMongoError("Update() Failed to Update. Check the inner error.", 1004, err)
	}
	fmt.Println("Updated ", result.ModifiedCount, " Document(s)")

	return nil
}

// Deletes a Mongo Document by its ID from the configured Mongo DB
// Example:
//
//	mongoDB := New(
//		WithDatabaseName("mydb"),
//		WithCollectionName("mycollection"),
//		WithURI("mongodb://localhost:27017"),
//	)
//	err := mongoDB.Delete(&MyMongoDocument{
//		FirstName: "John",
//		LastName:  "Doe",
//	}, "5e9b9b9b9b9b9b9b9b9b9b9b")
func (m *MongoDB) Delete(doc IMongoDocument, id string) error {
	collection, err := m.getCollection(doc)
	if err != nil {
		return err
	}
	objectID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		log.Fatal(err)
		return NewChuxMongoError("Delete() Failed to Get ObjectIDFromHex. Check the inner error.", 1005, err)
	}
	result, err := collection.DeleteOne(context.Background(), bson.M{"_id": objectID})
	if err != nil {
		log.Fatal(err)
		return NewChuxMongoError("Delete() Failed to Delete. Check the inner error.", 1005, err)
	}
	fmt.Println("Deleted ", result.DeletedCount, " Document(s)")

	return nil
}

// Returns a collection and db name from the IMongoDocument interface
// or the configured values if the interface is not implemented
func (m *MongoDB) getDBAndCollectionName(doc IMongoDocument) (string, string, error) {
	// Using the IMongoDocument interface, get the collection and database name. When the interface is implemented
	// the struct that implements it will have the option of overriding the configured collection and database name
	// with their implementation of the interface methods. This allows for a single struct to be used for multiple collections
	// Its a point of extensibility that is not needed by all use cases.

	var dbName string
	var collectionName string
	if len(doc.GetCollectionName()) > 0 {
		collectionName = doc.GetCollectionName()
	} else {
		collectionName = m.CollectionName
	}
	if len(doc.GetDatabaseName()) > 0 {
		dbName = doc.GetDatabaseName()
	} else {
		dbName = m.DatabaseName
	}

	if len(collectionName) == 0 || len(dbName) == 0 {
		return "", "", NewChuxMongoError("Did not create document. . Check the inner error for details", 1000, nil)
	}

	return collectionName, dbName, nil
}

// Returns the MongoDB collection from the IMongoDocument interface
func (m *MongoDB) getCollection(doc IMongoDocument) (*mongo.Collection, error) {
	client, err := m.Connect()
	if err != nil {
		return nil, err
	}
	collectionName, dbName, err := m.getDBAndCollectionName(doc)
	if err != nil {
		return nil, err
	}
	collection := client.Database(dbName).Collection(collectionName)
	if collection == nil {
		return nil, NewChuxMongoError(fmt.Sprintf("Unable to get the collection: %s from database: %s Check the inner error for details", collectionName, dbName), 1000, nil)
	}
	return collection, nil
}

func (m *MongoDB) CreateIndices(doc IMongoDocument, fieldNames ...string) (bool, error) {
	client, err := m.Connect()
	if err != nil {
		return false, err
	}
	collectionName, dbName, err := m.getDBAndCollectionName(doc)
	if err != nil {
		return false, errors.NewChuxDataStoreError("Unable to get the collection name and database name from the IMongoDocument interface. Check the inner error for details", err)
	}
	collection := client.Database(dbName).Collection(collectionName)
	if collection == nil {
		return false, NewChuxMongoError(fmt.Sprintf("Unable to get the collection: %s from database: %s Check the inner error for details", collectionName, dbName), 1000, nil)
	}
	for _, fieldName := range fieldNames {
		indexView := collection.Indexes()
		indexModel := mongo.IndexModel{
			Keys: bson.M{
				fieldName: 1,
			},
			Options: options.Index().SetUnique(true),
		}
		_, err := indexView.CreateOne(context.Background(), indexModel)
		if err != nil {
			return false, NewChuxMongoError(fmt.Sprintf("Unable to create the indicies: %s on collection: %s Check the inner error for details", fieldNames, collectionName), 1000, nil)
		}
	}
	return true, nil
}